### DISCUSSION TITLE
Structure member access of an array of objects causing over complex output

### DISCUSSION THREAD

**Discussion by Wall-AF** [NationalSecurityAgency/ghidra#7071] [Q&A]
I have an array of 10 fully defined `GlobalWaveDetails` objects instantiated in the data segment.  When Ghidra's decompiler produces C that references the first member of the structure, although not perfect, the output is okay-ish (see line 2 below).  However when referencing other structure members the array index becomes externlised and consequently is multiplied by the structures size and then added, making things less legible (see lines 3, 4 & 6 below):
```
         for (idx2 = 0; idx2 < 10; idx2 += 1) {
            if (((((GlobalWaveDetails *)g_deviceInfo_25a6 + idx2)->m_idDevice != 0) &&
                (*(int *)((int)&g_deviceInfo_25a6[0].m_inUse + idx2 * 8) != 0)) &&
               (bSuccess = IsWindow(*(HWND *)((int)&g_deviceInfo_25a6[0].m_hWnd + idx2 * 8)),
               bSuccess != FALSE)) {
               PostMessage(0,0,0x48c,*(HWND *)((int)&g_deviceInfo_25a6[0].m_hWnd + idx2 * 8));
               break;
            }
         }
```

Can anyone advise as to where this transformation is occurring (either/both of Ghidra's C and/or Java) so a fix could be found?

**Comment by DualTachyon** [comment]
Sorry to jump in, but check if these PRs help in anyway:

https://github.com/NationalSecurityAgency/ghidra/pull/6722
https://github.com/NationalSecurityAgency/ghidra/pull/6718

**Reply by Wall-AF** [reply]
> Sorry to jump in, but check if these PRs help in anyway:
> 
> #6722 #6718

Added both to just the decompler code, but no difference, unfortunately.  Thanks for the suggestions, keep 'em coming!

**Comment by f-raZ0R** [comment]
Bumping this discussion, as I've got stuff involving loops with arrays like this all over the codebase i'm analyzing, and it gets really hard to tell what's going on, or what is even inside of a struct
